internals_fsutils.go

```diff
@@ -382,53 +386,51 @@ func createZip(archiveName string, files map[string][]byte) error {
 		if err != nil {
 			return err
 		}
-		_, err = f.Write([]byte(fcont))
+		_, err = io.Copy(f, fcont)
 		if err != nil {
 			return err
 		}
 	}
 
-	// Make sure to check the error on Close.
-	err := w.Close()
-	if err != nil {
-		return err
-	}
-
-	err = ioutil.WriteFile(archiveName, buf.Bytes(), defaultFilePermissions)
-	if err != nil {
-		return err
-	}
-
-	return nil
+	// The Close call writes the checksum, so return the error
```

Comment:
I don't see how writing the checksum affects our defer pattern. If you don't want to leave a bad zip on disk, I suggest you handle that error case explicitly (demonstrated below).

I would rather defer closing the writer because we want to close `w` when `io.Copy` errors and we return prematurely.

I think a better pattern would be to name the return type `(err error)` and defer an anonymous function like so:
```
archiveFile, err := os.Create(archiveName)
if err != nil {
    return err
}
w := zip.NewWriter(archiveFile)
defer func() {
    if cerr := w.Close(); cerr != nil && err == nil {
        err = cerr
    }
    if cerr := archiveFile.Close(); cerr != nil && err == nil {
        err = cerr
    }
    if err != nil {
        // Attempt to delete incomplete archive
        // We can't do anything about a failure to remove and
        // the original error is more relevant to the caller
        os.Remove(archiveName)
    }
}
```